### PR TITLE
Add execution summary framework

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

Extracted from https://github.com/tikv/tikv/pull/3898

This PR adds optional execution summary framework for batch executors. Additionally, as an example, it also includes the change that makes batch table scan executor supporting execution summary.

## What are the type of the changes? (mandatory)

- New feature (change which adds functionality)

## How has this PR been tested? (mandatory)

Unit test in batch table scan executor. Note that execution summary is implemented in scan executor, which is shared in both table scan and index scan, so index scan executor also benefits from this change.

### CODE REVIEW

File: src/coprocessor/dag/batch_executor/statistics.rs [hunk 1]
```diff
@@ -24,17 +29,155 @@ pub struct BatchExecuteStatistics {
...
+/// Execution summaries to support `EXPLAIN ANALYZE` statements.
+#[derive(Default)]
+pub struct ExecSummary {
+    /// Total time cost in this executor.
+    pub time_processed_ms: Arc<atomic::AtomicUsize>,
+
+    /// How many rows this executor produced totally.
+    pub num_produced_rows: usize,
+
+    /// How many times executor's `next()` is called.
```
Comment (rleungx) [inline src/coprocessor/dag/batch_executor/statistics.rs]: `next_batch()`?

File: src/coprocessor/dag/batch_executor/statistics.rs [hunk 2]
```diff
@@ -24,17 +29,155 @@ pub struct BatchExecuteStatistics {
...
+    #[inline]
+    fn drop(&mut self) {
+        let elapsed = (self.start_time.elapsed_secs() * 1000f64) as usize;
+        self.collect_target
+            .fetch_add(elapsed, atomic::Ordering::Relaxed);
+    }
+}
+
+/// A normal `ExecSummaryCollector`. Acts like `collect = true`.
+pub struct ExecSummaryCollectorNormal {
```
Comment (rleungx) [inline src/coprocessor/dag/batch_executor/statistics.rs]: How about just using a boolean to decide whether we need `ExecSummaryCollector`? Or maybe use an `Option`?

File: src/coprocessor/dag/batch_executor/statistics.rs (hunk 1 above)
Comment (breeswish) [reply src/coprocessor/dag/batch_executor/statistics.rs]: Nice catch!

File: src/coprocessor/dag/batch_executor/statistics.rs (hunk 2 above)
Comment (breeswish) [reply src/coprocessor/dag/batch_executor/statistics.rs]: Boolean or Option will introduce runtime cost (to check the bool), while using this way there will be completely no cost when it is not needed.

Comment (rleungx) [reply src/coprocessor/dag/batch_executor/statistics.rs]: How about `ExecSummaryCollectorEnabled`?

File: src/coprocessor/dag/batch_executor/scan_executor.rs [hunk 3]
```diff
@@ -49,7 +49,11 @@ pub trait ScanExecutorImpl: Send {
     ) -> Result<()>;
 }
 
-pub struct ScanExecutor<S: Store, I: ScanExecutorImpl, P: PointRangePolicy> {
+pub struct ScanExecutor<C: ExecSummaryCollector, S: Store, I: ScanExecutorImpl, P: PointRangePolicy>
```
Comment (rleungx) [inline src/coprocessor/dag/batch_executor/scan_executor.rs]: We need to add some comments

File: src/coprocessor/dag/batch_executor/statistics.rs (hunk 2 above)
Comment (breeswish) [reply src/coprocessor/dag/batch_executor/statistics.rs]: That's ok

File: src/coprocessor/dag/batch_executor/scan_executor.rs [hunk 4]
```diff
@@ -206,11 +218,14 @@ impl<S: Store, I: ScanExecutorImpl, P: PointRangePolicy> BatchExecutor for ScanE
...
 
+        let timer = self.summary_collector.start_record_duration();
+        self.summary_collector.inc_iterations();
+
         let mut data = self.imp.build_column_vec(expect_rows);
         let is_drained = self.fill_column_vec(expect_rows, &mut data);
 
         data.assert_columns_equal_length();
-
+
```
Comment (rleungx) [inline src/coprocessor/dag/batch_executor/scan_executor.rs]: remove this?

File: src/coprocessor/dag/batch_executor/table_scan_executor.rs [hunk 5]
```diff
@@ -628,12 +641,91 @@ mod tests {
...
+                .num_produced_rows,
+            3
+        );
+        assert_eq!(
+            statistics.summary_per_executor[1]
+                .as_ref()
+                .unwrap()
+                .num_iterations,
+            2
+        );
```
Comment (AndreMouche) [inline src/coprocessor/dag/batch_executor/table_scan_executor.rs]: L690-L705 are asserting the statistics are the same as before? Repeated code exist, is there any way to remove it?

Comment (breeswish) [reply src/coprocessor/dag/batch_executor/table_scan_executor.rs]: Yes, it is exactly asserting the statistics are the same as before. This is the test purpose. When nothing happens, the statistics should be the same as before.

**Pull request merged**